Content.Shared/Teleportation/Systems/SharedPortalSystem.cs

```diff
@@ -145,7 +151,102 @@ private void OnCollide(EntityUid uid, PortalComponent component, ref StartCollid
 
         // no linked entity--teleport randomly
         if (component.RandomTeleport)
-            TeleportRandomly(uid, subject, component);
+            TeleportRandomly(portal, subject, component);
+    }
+
+    private void OnCanDragDropOn(EntityUid uid, PortalComponent component, ref CanDropTargetEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        args.CanDrop = CanInsert(uid, component, args.Dragged);
+        args.Handled = true;
+    }
+
+    private bool TryInsert(EntityUid uid, EntityUid toInsertId, EntityUid? userId, PortalComponent? component = null)
+    {
+        if (!Resolve(uid, ref component))
+            return false;
+
+        if (!CanInsert(uid, component, toInsertId))
+            return false;
+
+        bool insertingSelf = userId == toInsertId;
+
+        var delay = insertingSelf ? component.EntryDelay : component.DraggedEntryDelay;
+
+        if (userId != null && !insertingSelf)
+            _popup.PopupEntity(Loc.GetString("portal-being-inserted", ("user", Identity.Entity((EntityUid) userId, EntityManager))), toInsertId, toInsertId, PopupType.Large);
+
+        if (delay <= 0 || userId == null)
+        {
+            EnterPortal(uid, toInsertId, component);
+            return true;
+        }
+
+        // Can't check if our target AND disposals moves currently so we'll just check target.
+        // if you really want to check if disposals moves then add a predicate.
+        var doAfterArgs = new DoAfterArgs(EntityManager, userId.Value, delay, new PortalDoAfterEvent(), uid, target: toInsertId, used: uid)
+        {
+            BreakOnDamage = true,
+            BreakOnMove = true,
+            NeedHand = false
+        };
+
+        _doAfterSystem.TryStartDoAfter(doAfterArgs);
+        return true;
+    }
+
+
+    private void OnDoAfter(EntityUid uid, PortalComponent component, DoAfterEvent args)
+    {
+        if (args.Handled || args.Cancelled || args.Args.Target == null || args.Args.Used == null)
+            return;
+
+        EnterPortal(uid, args.Args.Target.Value, component);
+
+        args.Handled = true;
+    }
+
+    private bool CanInsert(EntityUid uid, PortalComponent component, EntityUid entity)
+    {
+        if (!component.TeleportOnDragDrop)
+            return false;
+        if (!Transform(uid).Anchored)
+            return false;
+
+        var storable = HasComp<ItemComponent>(entity);
+        if (!storable && !HasComp<BodyComponent>(entity))
+            return false;
+
```

Comment:
Don't a whole lot more entities than this insert into portals? I'm suss of this method in general.

Reply:
I assumed it would work similarly to disposal units so I just copied the code from there.

Reply:
Do you think I should remove this whole method?

Reply:
Yes, don't just copy-paste code without understanding it.